File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left_to_right_products = [nums[0] for _ in range(len(nums))]
        right_to_left_products = [nums[-1] for _ in range(len(nums))]

        for i in range(1, len(nums)):
            left_to_right_products[i] = left_to_right_products[i - 1] * nums[i]
        
        for i in range(len(nums) - 2, -1, -1):
            right_to_left_products[i] = right_to_left_products[i + 1] * nums[i]

        res = [0 for _ in range(len(nums))]
        for i in range(len(nums)):
            if i == 0:
                res[i] = right_to_left_products[i + 1]
            elif i == len(nums) - 1:
                res[i] = left_to_right_products[i - 1]
            else:
                res[i] = left_to_right_products[i - 1] * right_to_left_products[i + 1]

        return res

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] prefix = new int[nums.length];
        int[] postfix = new int[nums.length];

        // prefix
        for (int i = 0; i < nums.length; i++) {
            if (i == 0) {
                prefix[i] = nums[i];
            } else {
                prefix[i] = nums[i] * prefix[i - 1]; 
            }
        }

        // postfix
        for (int i = nums.length - 1; i >= 0; i--) {
            if (i == nums.length - 1) {
                postfix[i] = nums[i];
            } else {
                postfix[i] = nums[i] * postfix[i + 1];
            }
        }

        int[] res = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            if (i == 0) {
                res[i] = postfix[i + 1];
            } else if (i == nums.length - 1) {
                res[i] = prefix[i - 1];
            } else {
                res[i] = prefix[i - 1] * postfix[i + 1];
            }
        }

        return res;
    }
}